components/payments/internal/connectors/plugins/public/generic/accounts.go

```diff
@@ -0,0 +1,88 @@
+package generic
+
+import (
+	"context"
+	"encoding/json"
+	"time"
+
+	"github.com/formancehq/payments/internal/models"
+)
+
+type accountsState struct {
+	LastCreatedAtFrom time.Time `json:"lastCreatedAtFrom"`
+}
+
+func (p Plugin) fetchNextAccounts(ctx context.Context, req models.FetchNextAccountsRequest) (models.FetchNextAccountsResponse, error) {
+	var oldState accountsState
+	if req.State != nil {
+		if err := json.Unmarshal(req.State, &oldState); err != nil {
+			return models.FetchNextAccountsResponse{}, err
+		}
+	}
+
+	newState := accountsState{
+		LastCreatedAtFrom: oldState.LastCreatedAtFrom,
+	}
+
+	var accounts []models.PSPAccount
+	hasMore := false
+
+	for page := 0; ; page++ {
+		pagedAccounts, err := p.client.ListAccounts(ctx, int64(page), int64(req.PageSize), oldState.LastCreatedAtFrom)
+		if err != nil {
+			return models.FetchNextAccountsResponse{}, err
+		}
+
+		if len(pagedAccounts) == 0 {
+			break
+		}
+
+		for _, account := range pagedAccounts {
+			switch account.CreatedAt.Compare(oldState.LastCreatedAtFrom) {
+			case -1, 0:
+				// Account already ingested, skip
+				continue
+			default:
+			}
+
+			raw, err := json.Marshal(account)
+			if err != nil {
+				return models.FetchNextAccountsResponse{}, err
+			}
+
+			accounts = append(accounts, models.PSPAccount{
+				Reference: account.Id,
+				CreatedAt: account.CreatedAt,
+				Name:      &account.AccountName,
+				Metadata:  account.Metadata,
+				Raw:       raw,
+			})
+
+			newState.LastCreatedAtFrom = account.CreatedAt
+
+			if len(accounts) >= req.PageSize {
+				break
```

Comment:
https://github.com/formancehq/stack/pull/1711/files/b9838ade3dc4a5947d1e2dba77c19f16bd6dd228#diff-a46026682ab0d5879ea4636201f453e69cb89c45eb154e007d476c7f69716c0eR72

A few lines after, i'm breaking the outer loop !

Reply:
I think it's a bit fragile and easy to introduce bugs if we need to rely on two separate conditions in different for loops to control the exit of the main loop. Would be better to use for loop tags and do something like `break OUTER` since the later page size checks can also be skipped if we already have the maximum number of accounts.

When it comes to a situation where we need to introduce a nested for loop, I usually recommend trying to compartmentalise the inner loop in a subroutine to limit the interactions between the inner and outer loops.

Something like this where the inner loop is now only concerned with processing each account and doesn't need to know about the logic of the outer loop. The outer loop controls the inner loops access to data needing to be processed by avoiding passing any accounts we don't need in this result set.
```
for {
  needed = req.PageSize - len(accounts)
  if needed < 1 {
    break
  }

  pagedAccounts := client.Fetch(req.PageSize)
  
  if len(pagedAccounts) > needed {
    hasMore = true
  }

  results = loopOverPaged(pagedAccounts[0:needed])
  accounts = append(accounts, results...)
}
```